File: src/matlab/com/springbok/pattern/PatternENOR207V01Test.m
classdef PatternENOR207V01Test < TestUtility
% Tests methods of PatternENOR207V01 class.
  
% Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved.

  properties (Constant = true)
    
    Phi_input = [0.2; 4; 150];
    
    self = struct( ...
        'GainMax', 50.7);
    
    G_expected = [ ...
        49.008000000000003
        13.948500216800943
        -10.000000000000000
                 ];
    Gx_expected = [ ...
        49.008000000000003
        13.948500216800943
        -10.000000000000000
                  ];
    
  end % properties (Constant = true)
  
  methods
    
    function this = PatternENOR207V01Test(logFId, testMode)
    % Constructs a PatternENOR207V01Test
    %
    % Parameters:
    %   logFId - Log file identifier
    %   testMode - Test mode, if 'interactive' then beeps and pauses
      
    % Invoke superclass constructor
      if nargin == 0
        superArgs = {};
        
      else
        superArgs{1} = logFId;
        superArgs{2} = testMode;
        
      end % if
      this@TestUtility(superArgs{:});
      
    end % PatternENOR207V01Test()
    
    function test_PatternENOR207V01(this)
    % Tests PatternENOR207V01 method.
      
      p = PatternENOR207V01();
      
      this.assert_true( ...
          'PatternENOR207V01', ...
          'PatternENOR207V01', ...
          this.HIGH_PRECISION_DESC, ...
          PatternUtility.compare_pattern(this.self, p));
      
    end % test_PatternENOR207V01()
    
    function test_gain(this)
    % Tests gain method.
      
      warning('off');
      
      p = PatternENOR207V01();
      
      [G_actual, Gx_actual] = p.gain(this.Phi_input);
      
      t(1) = max(abs(G_actual - this.G_expected) < this.HIGH_PRECISION);
      t(2) = max(abs(Gx_actual - this.Gx_expected) < this.HIGH_PRECISION);
      
      this.assert_true( ...
          'PatternENOR207V01', ...
          'gain', ...
          this.HIGH_PRECISION_DESC, ...
          min(t));
      
      [G, Gx, fH] = p.gain([0:.001:180], 'PlotFlag', true);
      pause(1); close(fH);
      
      warning('on');
      
    end % test_gain()
    
  end % methods
  
end % classdef

File: src/java/com/springbok/pattern/PatternENOR207V01Test.java
/* Copyright (C) 2022 Springbok LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.springbok.pattern;

import com.springbok.utility.TestUtility;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertTrue;

/**
 * Tests methods of PatternENOR207V01 class.
 */
public class PatternENOR207V01Test {
	private static final double[] Phi_input = new double[] { 0.2, 4.0, 150.0 };
	private static final double[] G_expected = new double[] { 49.008000000000003, 13.948500216800943, -10.0 };
	private static final double[] Gx_expected = new double[] { 49.008000000000003, 13.948500216800943, -10.0 };

	/**
	 * Tests PatternENOR207V01 method.
	 */
	@Test
	public void test_PatternENOR207V01() {
		PatternENOR207V01 p = new PatternENOR207V01();

		assertTrue(TestUtility.isDoublesEquals(50.7, p.getGainMax()));
	}

	/**
	 * Tests gain method.
	 */
	@Test
	public void test_gain() {
		PatternENOR207V01 p = new PatternENOR207V01();

		Gain[] gains = new Gain[Phi_input.length];
		Map<String, Object> options = new HashMap<String, Object>();
		options.put("DoValidate", false);

		for (int i = 0; i < gains.length; i++) {
			Gain g = p.gain(Phi_input[i], options);
			gains[i] = g;
		}
		for (int i = 0; i < gains.length; i++) {
			assertTrue(TestUtility.isDoublesEquals(G_expected[i], gains[i].G));
			assertTrue(TestUtility.isDoublesEquals(Gx_expected[i], gains[i].Gx));
		}
	}
}
